x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/WatcherLifeCycleService.java

```diff
@@ -96,6 +110,39 @@ public void clusterChanged(ClusterChangedEvent event) {
             return;
         }
 
+        if (event.localNodeMaster() && checkForLegacyTemplates.get()) {
+            List<String> existingTemplatesToDelete = new ArrayList<>(LEGACY_WATCHER_INDEX_TEMPLATES.size());
+            ImmutableOpenMap<String, IndexTemplateMetadata> clusterLegacyTemplates = event.state().getMetadata().templates();
+            for (String legacyWatcherIndexTemplate : LEGACY_WATCHER_INDEX_TEMPLATES) {
+                if (clusterLegacyTemplates.containsKey(legacyWatcherIndexTemplate)) {
+                    existingTemplatesToDelete.add(legacyWatcherIndexTemplate);
+                }
+            }
+
+            if (existingTemplatesToDelete.isEmpty() == false) {
+                // if someone else is executing the deletion of templates (due to fast successive cluster updates), we'll skip doing so
+                if (legacyTemplatesDeleteInProgress.compareAndSet(false, true) == false) {
+                    return;
+                }
+
+                MetadataIndexTemplateService.removeTemplates(clusterService, LEGACY_WATCHER_INDEX_TEMPLATES, ActionListener.wrap(r -> {
+                    legacyTemplatesDeleteInProgress.set(false);
+                    // we've done it so we shouldn't check anymore
+                    checkForLegacyTemplates.set(false);
+                    logger.debug("deleted legacy Watcher index templates [{}]", String.join(",", LEGACY_WATCHER_INDEX_TEMPLATES));
+                }, e -> {
+                    legacyTemplatesDeleteInProgress.set(false);
+                    logger.debug(
+                        new ParameterizedMessage(
+                            "unable to delete legacy Watcher index templates [{}]",
+                            String.join(",", LEGACY_WATCHER_INDEX_TEMPLATES)
+                        ),
+                        e
+                    );
+                }));
+            }
```

Comment:
Should `checkForLegacyTemplates` be set to false when `existingTemplatesToDelete` list is empty?